stan/math/rev/fun/add.hpp

```diff
@@ -0,0 +1,538 @@
+        = (Ad.array() + cd).matrix().eval().unaryExpr([](double x) {
+            return new vari(x, false);
+          });
+  }
+
+  virtual void chain() {
+    using Eigen::Map;
+    matrix_d adjAplusc(rows_, cols_);
+    adjAplusc = Map<matrix_vi>(variRefAplusc_, rows_, cols_).adj();
+    Map<matrix_vi>(variRefA_, rows_, cols_).adj() += adjAplusc;
+  }
+};
+
+/**
+ * This is a subclass of the vari class for matrix
+ * addition A + c, where A is a N by M matrix of
+ * doubles c is a scalar var.
+ *
+ * The class stores the structure of each matrix,
+ * the double values of A and c, and pointers to
+ * the varis for A and c if A or c is a var. It
+ * also instatiates and stores pointers to varis
+ * for all elements of A + c.
+ *
+ * @tparam Ta type of elements in matrix A
+ * @tparam Tc type of elements in matrix B
+ * @tparam R number of rows (common to A and B)
+ * @tparam C number of columns (common to A and B)
+ */
+
+template <typename Tc, int R, int C>
+class add_mat_scal_vari<double, Tc, R, C> : public vari {
+ public:
+  int rows_;
+  int cols_;
+  int size_;
+  double* Ad_;
+  double cd_;
+  vari* variRefc_;
+  vari** variRefAplusc_;
+
+  /**
+   * Constructor for add_mat_vari.
+   *
+   * All memory is allocated in ChainableStack's
+   * stack_alloc arena.
+   *
+   * It is critical for the efficiency of this object
+   * that the constructor create new varis that aren't
+   * popped onto the var_stack_, but rather are
+   * popped onto the var_nochain_stack_. This is
+   * controlled by the second argument to
+   * vari's constructor.
+   *
+   * @param A matrix
+   * @param c scalar
+   **/
+  add_mat_scal_vari(const Eigen::Matrix<double, R, C>& A, const Tc c)
+      : vari(0.0),
+        rows_(A.rows()),
+        cols_(A.cols()),
+        size_(A.size()),
+        Ad_(ChainableStack::instance_->memalloc_.alloc_array<double>(size_)),
+        cd_(c.val()),
+        variRefc_(c.vi_),
+        variRefAplusc_(
+            ChainableStack::instance_->memalloc_.alloc_array<vari*>(size_)) {
+    using Eigen::Map;
+
+    double cd = c.val();
+    Map<matrix_d> Ad(Ad_, rows_, cols_);
+    Ad = A.val();
+    Map<matrix_vi>(variRefAplusc_, rows_, cols_)
+        = (Ad.array() + cd).matrix().unaryExpr([](double x) {
+            return new vari(x, false);
+          });
+  }
+
+  virtual void chain() {
+    using Eigen::Map;
+    matrix_d adjAplusc(rows_, cols_);
+    adjAplusc = Map<matrix_vi>(variRefAplusc_, rows_, cols_).adj();
+    variRefc_->adj_ += adjAplusc.sum();
+  }
+};
+
+/**
+ * Return the sum of the specified matrix and specified scalar.
+ *
+ * @tparam Mat type of the matrix or expression
+ * @tparam Scal type of the scalar
+ * @param m Matrix or expression.
+ * @param c Scalar.
+ * @return The matrix plus the scalar.
+ */
+template <typename Mat, typename Scal, require_eigen_t<Mat>* = nullptr,
+          require_stan_scalar_t<Scal>* = nullptr,
+          require_any_var_t<typename value_type<Mat>::type, Scal>* = nullptr>
```

Comment:
```suggestion
template <typename Mat, typename Scal, typename Ta = value_type_t<Mat>,
          require_eigen_t<Mat>* = nullptr,
          require_stan_scalar_t<Scal>* = nullptr,
          require_any_var_t<Ta, Scal>* = nullptr>
```

Then you don't need the additional typedef for ```Ta``` below